Custom modifier equivalence, duplication, appending and inflation

lambdageek opened the pull request:
This all stems from ECMA 335, 7.1.1:

> The CLI itself shall treat required and optional modifiers in the same
> manner. Two signatures that differ only by the addition of a custom
> modifier (required or optional) shall not be considered to match. Custom
> modifiers have no other effect on the operation of the VES.

This rule means that when comparing signatures, we have to treat two types that
differ only in their custom modifiers as distinct. In particular, we have to do
this when comparing signatures in order to fill in the vtable for a class.  And
moreover, we have to concatenate custom modifiers when doing generic
instantiation.  If a method in a generic type has signature `T modopt(IsConst)
Method()` and we instantiate T with `int32 modopt(IsLong)`, we are expected to
produce the signature `int32 modopt(IsConst) modopt(IsLong) Method()` (with the
custom modifiers on `int32` in that order).

So this PR implements the following operations:
1. custom modifier appending during `mono_metadata_type_dup_with_cmods`
2. custom modifier-respecting comparison in `do_mono_metadata_type_equal`

To do this, we actually have to introduce a second representation of custom
modifiers: in the normal representation, all the cmod type specs come from a
single image.  Whereas now, due to inflation of signatures, we can have a mix
of modifiers from different images.  So in the MonoAggregateModContainer
representation, we pair each typespec token with the image in which it is to be
resolved.

We keep the old representation (a single image to resolve the entire set of
typespec tokens) because it's natural when you're just reading from a single
image.  Also this representation is part of the embedding API.

---

### But that's not all ###


It is possible for generic parameters of a generic class to appear in the
_custom modifiers_ of a type in a method signature.  For example:
```
    T modopt(Nullable`1<T>) Method ()
```

In this case, when we inflate this signature, we need to inflate not only the
return type, but also the modifiers on a return type.

That means our old aggregate representation (where each modifier is a pair of a
MonoImage and a typespec token) is inadequate: we can inflate a signature and
get a custom modifier that doesn't appear as a typespec in any table of any
existing image.

This is the same situation as generic instances: as a program runs, it creates
generic instances that are not part of the source code.  So we solve it the
same way as with generic instances.

Our new `MonoAggregateModContainer` representation is a pointer into memory owned
by a MonoImageSet that contains an entire MonoType for each custom modifier.

The invariant is that the aggregate modifier container is allocated from a
mempool of a `MonoImageSet` that includes the images of all the modifiers.

In general, when inflating, we first create a provisional candidate
MonoAggregateModContainer on the stack and then look it up the image set that
should own it and then look for a "canonical" mod container in the
`MonoImageSet:aggergate_modifiers_cache`

---

Also had to bump the AOT image format, which now always reads back in using the aggregate representation.

---

This should address #10564, #10565, #12422, #10945 (again), #6936, and the parts of #11350 that **don't** deal with arrays.

---

TODO:

- [x] Add some representative test cases to this PR.
- [ ] Modify our acceptance tests to run the various disabled Roslyn tests.

lambdageek commented:
We also pass [`ModifiersWithConstructedType_01`](https://github.com/dotnet/roslyn/blob/defeca5d0e44915fc978cfd93d3de2e4a6d0dbf4/src/Compilers/CSharp/Test/Symbol/Symbols/CustomModifiersTests.cs#L2044) from the Roslyn test suite.  So we should make a PR against roslyn to enable that test.

lambdageek commented:
I'm doing a lot of cloning -  I would especially welcome sanity checking of all the memory management.

lambdageek commented:
@monojenkins build failed

jaykrell commented on mono/metadata/class.c:
```diff
@@ -632,18 +635,40 @@ can_inflate_gparam_with (MonoGenericParam *gparam, MonoType *type)
+				return NULL;
+			else
+				return type;
+		}
```
`return changed ? type : NULL;`?

jaykrell commented on mono/metadata/class.c:
```diff
@@ -688,16 +717,40 @@ inflate_generic_type (MonoImage *image, MonoType *type, MonoGenericContext *cont
+			char *tname;
+			char *vname;
+			tname = mono_type_full_name (type);
+			vname = mono_type_full_name (inst->type_argv[num]);
```
`char* vname = vname = mono_type_full_name (inst->type_argv[num]);` etc.?

jaykrell commented on mono/metadata/class.c:
```diff
@@ -760,18 +831,82 @@ inflate_generic_type (MonoImage *image, MonoType *type, MonoGenericContext *cont
+	int count = mono_type_custom_modifier_count (type);
+	gboolean changed = FALSE;
+	MonoType *new_type = g_alloca (mono_sizeof_type_with_mods (count, TRUE));
+	memcpy (new_type, type, MONO_SIZEOF_TYPE);
```
`*new_type = *type`?

kg replied:
Previously it needed to be an alloc + some copies because mono_types with cmods are bigger than sizeof(type)

lambdageek replied:
yea this line probably merits a comment or a helper function - I am specifically just copying the "plain type" slice of `type` into `new_type`.  It's true that `*new_type = *type` would do the same thing, but it's perhaps even less evident than a memcpy.

jaykrell replied:
It should be obvious, at least in C.

jaykrell commented:
Wow.

lambdageek commented:
Also reversed the order that we store custom modifiers.
If you write IL like this:
```
   int32 modopt (Foo) modopt (Bar)
```
both .NET Framework ilasm and Mono's ilasm will produce an assembly with this:
```
   cmod_opt [encoding of Bar] cmod_opt [encoding of Foo] I4
```
So when parsing, we should store it with Foo first, then Bar.

And `monodis` should display it in the same order that `ilasm` consumed it in.

lambdageek commented:
@monojenkins build failed

alexanderkyte commented on mono/metadata/class.c:
```diff
@@ -688,16 +717,38 @@ inflate_generic_type (MonoImage *image, MonoType *type, MonoGenericContext *cont
 			return NULL;			
 		}
 
+#ifdef DEBUG_INFLATE_CMODS
```
Can we hook this up to some autogen setting or document where it's set? Does someone just add a #define somewhere? Could you add a commented-out copy of that define, so people know where to put it?

lambdageek replied:
There's an `#undef DEBUG_INFLATE_CMODS` at the top of the file:
https://github.com/mono/mono/pull/13320/files#diff-f085caaf39e861140ff40b9ef0c516d9R76

alexanderkyte commented on mono/metadata/class.c:
```diff
@@ -760,18 +829,85 @@ inflate_generic_type (MonoImage *image, MonoType *type, MonoGenericContext *cont
+	}
+#endif
+	if (changed) {
+		MonoType *new_type = g_alloca (mono_sizeof_type_with_mods (count, TRUE));
```
Can you add some assertions around this type of allocation (around the maximum return value of "mono_sizeof_type_with_mods") so that if this explodes the alt stack or something, it won't be a big debugging session?

alexanderkyte commented on mono/metadata/icall.c:
```diff
@@ -7928,9 +7930,12 @@ type_array_from_modifiers (MonoImage *image, MonoType *type, int optional, MonoE
-		if ((optional && !cmods->modifiers [i].required) || (!optional && cmods->modifiers [i].required)) {
-			if (!add_modifier_to_array (domain, image, &cmods->modifiers [i], res, count , error))
+	for (i = cmod_count - 1; i >= 0; --i) {
+		gboolean required;
```
How does this code get hit/tested? Might be good to document. It seems difficult to go from the type of breakage you'd see if this misbehaved back to this code.

alexanderkyte commented on mono/metadata/metadata-internals.h:
```diff
@@ -36,14 +36,86 @@ struct _MonoType {
+ *
+ * Because the types come from different images, we allocate the aggregate
+ * custom modifiers container object in the mempool of a MonoImageSet to ensure
+ * that it doesn't have dangling image pointers.
```
Well-written. 👍

alexanderkyte commented on mono/metadata/metadata.c:
```diff
@@ -2426,7 +2430,40 @@ dump_ginst (MonoGenericInst *ginst)
 }*/
 
+static gboolean
+aggregate_modifiers_equal (gconstpointer ka, gconstpointer kb)
```
At some point all of this modifier stuff probably deserves it's own file, outside of metadata. The churn might be too much for this PR though.

alexanderkyte commented on mono/metadata/reflection.c:
```diff
@@ -459,6 +459,12 @@ mono_type_get_object_checked (MonoDomain *domain, MonoType *type, MonoError *err
+	// world since they're hard to distinguish from plain types using the
+	// reflection APIs, but they are not ReferenceEqual to the unadorned
+	// types.
+	g_assert (!type->has_cmods);
```
Is this a "temporarily we don't support this" or a "this will never work" scenario?

lambdageek replied:
It's a "this is a bug in the runtime if it ever happens" scenario.

alexanderkyte commented:
Left a few comments. Pretty readable all-in-all. I would feel better about it as a reviewer if I knew we had more coverage of it. Any chance we know of a larger suite testing this in one of the other dotnet foundation repos?

vargaz commented:
Wouldn't be better to only use the new representation, and do some kind of conversion for the embedding api ?

lambdageek commented:
> Wouldn't be better to only use the new representation, and do some kind of conversion for the
> embedding api ?

Yea, that's possible.  I thought there may be some benefit to keeping the old representation - it's a bit more compact.

lambdageek commented:
Addressed @alexanderkyte comments.

Added a new reflection test.

Fixed a bug where our SRE was emitting cmods in reverse order.  Undid the order-reversing from #11308.  The same test from that PR still works - in effect we shifted the cmod reversal from reflection query (the wrong place for it) to SRE (the right place for it).

lambdageek commented:
@vargaz on second thought, getting rid fo the old custom mod representation is tricky: it will complicate the medatada parsing code and possibly make us resolve types too early.

The old representation is just an image and an array of tokens and it's allocated with the MonoType.  So there's no extra memory to manage, and we don't have to resolve any of the types - it matches up pretty nicely with what's in the assemblies.

The new representation is a pool-allocated array of `MonoType*`.  So we'd have to turn each image+token into a type right in the metadata parser.  So things can go wrong earlier.  Which might be more difficult to ignore or to recover from.

I would prefer to leave things with the two representations, if we can.
